File: rSVParser/src/ports.rs
use std::vec::Vec;

use sv_parser::{unwrap_node, SyntaxTree, RefNode};

use crate::get_identifier;

/// Traverses the tree from a node and returns a vector of all port declarations.
/// Works for modules, however it needs a small tweak to work with interfaces as well.
pub fn get_ansi_ports(tree: &SyntaxTree, module: RefNode) -> Vec<[String; 6]> {
    let mut ports = Vec::new();
    for node in module {
        match node {
            RefNode::AnsiPortDeclaration(port) => {
                let direction = match unwrap_node!(port, PortDirection) {
                    Some(RefNode::PortDirection(x)) => {
                        String::from(tree.get_str(x).unwrap().trim())
                    },
                    // Can occur when using interfaces for example
                    _ => String::new(),
                };

                let net_type = match unwrap_node!(port, NetType, DataType) {
                    Some(RefNode::NetType(x)) => String::from(tree.get_str(x).unwrap().trim()),
                    Some(RefNode::DataType(x)) => {
                        // Split out first word. Some data types here get screwy and I don't
                        // feel like traversing the entire tree
                        String::from(tree.get_str(x).unwrap().split(" ").next().unwrap())
                    },
                    _ => String::new(),
                };

                let dimension = match unwrap_node!(port, PackedDimension) {
                    Some(RefNode::PackedDimension(x)) => String::from(tree.get_str(x).unwrap().trim()),
                    _ => String::new(),
                };

                let name = unwrap_node!(port, PortIdentifier).unwrap();
                let name = get_identifier(name).unwrap();
                let name = String::from(tree.get_str(&name).unwrap());

                ports.push([
                    direction,
                    net_type,
                    String::new(),
                    dimension,
                    name,
                    String::new(),
                ]);
            }
            _ => continue,
        }
    }

    return ports;
}

pub fn get_nonansi_ports(tree: &SyntaxTree, module: RefNode) -> Vec<[String; 6]> {
    let mut ports = Vec::new();
    for node in module {
        match node {
            RefNode::PortDeclaration(port) => {
                let direction = match unwrap_node!(port, PortDirection) {
                    Some(RefNode::PortDirection(x)) => {
                        String::from(tree.get_str(x).unwrap().trim())
                    },
                    // Can occur when using interfaces for example
                    _ => String::new(),
                };

                let net_type = match unwrap_node!(port, NetType, DataType) {
                    Some(RefNode::NetType(x)) => String::from(tree.get_str(x).unwrap().trim()),
                    Some(RefNode::DataType(x)) => {
                        // Split out first word. Some data types here get screwy and I don't
                        // feel like traversing the entire tree
                        String::from(tree.get_str(x).unwrap().split(" ").next().unwrap())
                    },
                    _ => String::new(),
                };

                let dimension = match unwrap_node!(port, PackedDimension) {
                    Some(RefNode::PackedDimension(x)) => String::from(tree.get_str(x).unwrap().trim()),
                    _ => String::new(),
                };

                let name = unwrap_node!(port, PortIdentifier).unwrap();
                let name = get_identifier(name).unwrap();
                let name = String::from(tree.get_str(&name).unwrap());

                ports.push([
                    direction,
                    net_type,
                    String::new(),
                    dimension,
                    name,
                    String::new(),
                ]);
            }
            _ => continue,
        }
    }

    return ports;
}

File: rSVParser/sv_hierarchy/ports.py
from typing import List, Tuple

from py_sv_parser import SyntaxTree, SyntaxNode, unwrap_node, unwrap_locate

from .helpers import get_str_or_default

def get_ports(
    tree: SyntaxTree,
    declaration: SyntaxNode
) -> List[Tuple]:
    """Pull all ANSI port declarations"""
    ports = []
    for node in declaration:
        if node.type_name == "AnsiPortDeclaration":
            direction = unwrap_node(node, ["PortDirection"])
            if direction is not None:
                direction = tree.get_str(direction).strip()
            else:
                direction = ""

            dimension = unwrap_node(node, ["PackedDimension"])
            if dimension is not None:
                dimension = tree.get_str(dimension).strip()
            else:
                dimension = ""

            name = unwrap_node(node, ["PortIdentifier"])
            name = unwrap_locate(name)
            name = tree.get_str(name)

            net_type = unwrap_node(node, ["NetType", "DataType"])
            if net_type is None:
                net_type = ""
            elif net_type.type_name == "NetType":
                net_type = tree.get_str(net_type).strip()
            elif net_type.type_name == "DataType":
                net_type = tree.get_str(net_type).split(" ")[0]

            ports.append((
                direction,
                net_type,
                "",
                dimension,
                name,
                ""
            ))

    return ports
